### ISSUE TITLE
The fltk::dialog::message window has no title

### ISSUE THREAD

**Issue by mark-summerfield** [fltk-rs/fltk-rs#1006]
The fltk::dialog::message window has no title, nor does the API allow the user to set a title.
Ideally the user should be able to specify a title as well as the message text.

https://docs.rs/fltk/1.2.17/fltk/dialog/fn.message.html

**Comment by MoAlyousef** [comment]
These just call fl_message(…) which doesn’t have a title field. FLTK allows customizing the dialog’s title via another function, fl_message_title which isn’t currently exposed in fltk-rs, but I’ll add it this evening. The api should look like:
```rust
dialog::message_title("some title");
dialog::message(args…); // works for alert, choice etc. 
```
There’s also an fl_message_default_title(), which allows setting the title only once. 

Ideally though, especially for apps with their own styling, a user defined dialog window should be preferable. Something like:
https://gist.github.com/MoAlyousef/ce50902cb7929760b89af60b3702fc1d

**Comment by mark-summerfield** [comment]
I tried the example on gist that you gave.

1. When I run it on Linux, if I click "Click" and enter the text say, "aaa" and then after clicking "Ok" the "aaa" shows up correctly in the frame. However, if I then repeat the process but with new text, say "WWW", the frame shows the new text _on top of_ the old text! This is even more apparent if you do it a third time with longer text.

2. Usually when I show a modal dialog like this I like to delete it after use, but this one is only hidden. Is it possible to delete when ok is clicked? I tried changing the ok.set_callback() to use win.delete(), win::<WidgetBase>.delete() and app::delete_widget(win), but all of them produced errors.

Thanks for your help.

**Comment by MoAlyousef** [comment]
I've added dialog::message_title() and message_title_default(). 

1- I've also fixed the gist, it was because of the frame's FrameType (which I've changed, another option was to call redraw in the button's callback).

2- Closing an FLTK window destroys it and its children. When you call show_dialog() (in the gist) again on it, it respawns a new window and new children.

**Comment by mark-summerfield** [comment]
I've now switched to using the lastest rust-rs using git in my Cargo.toml so I've tried dialog::message_title_default() and it works great. (I'll also try message_title().)

I applied your changes to my version of your dialog example and now it all works fine.

For modal dialogs I like to create/pop-up/retrieve OK/Cancel and any data/delete.
But unfortunately I cannot find a way to delete. The delete must be done in the callback to allow for the retrieval of data before deletion, but I can't get it to work:

```
    btn.set_callback(move |_| {
        let dlg = show_dialog();
        frame.set_label(&dlg.value());
        //app::delete_widget(dlg); 
        //window::Window::delete(dlg);
    });
```
The first commented out line if uncommented produces:
`the trait `fltk::prelude::WidgetBase` is not implemented for `MyDialog``,
and the second produces:
`expected struct `DoubleWindow`, found struct `MyDialog``

Thanks for your help!

**Comment by MoAlyousef** [comment]
You can expose the window as a field of the MyDialog:
```rust
pub struct MyDialog {
    win: window::Window,
    inp: input::Input,
}
```

Then you can call either `window::Window::delete(dlg.win); or app::delete_widget(dlg.win);`

Another option is to implement the Drop trait on the MyDialog, and call the deleter there.

**Comment by mark-summerfield** [comment]
I followed your advice adding the window to the MyDialog struct and I used app::delete_widget(dlg.win) in the callback. Thank you.

I think you've now changed fltk-rs's behavior. The dialog now seems to have the parent window's title (label) "My Dialog". However neither dialog::message_title() nor dialog::message_title_default() seem to be able to override it.

I've attached my edited main.rs which I had to gzip to make it work on github
[main.rs.gz](https://github.com/fltk-rs/fltk-rs/files/7604290/main.rs.gz)
.

**Comment by MoAlyousef** [comment]
This is a custom dialog, so it wouldn't respond to dialog::message_title(). You can specify whatever you want in the constructor and show_dialog function:
```rust
use fltk::{
    app, button,
    enums::{Color, Font, FrameType},
    frame, group, input,
    prelude::*,
    window,
};

fn style_button(btn: &mut button::Button) {
    btn.set_color(Color::Cyan);
    btn.set_frame(FrameType::RFlatBox);
    btn.clear_visible_focus();
}

pub fn show_dialog(title: &str) -> MyDialog {
    MyDialog::new(title)
}

pub struct MyDialog {
    window: window::Window,
    input: input::Input,
}

impl MyDialog {
    pub fn new(title: &str) -> Self {
        let mut win = window::Window::default()
            .with_size(400, 100)
            .with_label(title);
        win.set_color(Color::from_rgb(240, 240, 240));
        let mut pack = group::Pack::default()
            .with_size(300, 30)
            .center_of_parent()
            .with_type(group::PackType::Horizontal);
        pack.set_spacing(20);
        frame::Frame::default()
            .with_size(80, 0)
            .with_label("Enter name:");
        let mut input = input::Input::default().with_size(100, 0);
        input.set_frame(FrameType::FlatBox);
        let mut ok =
            button::Button::default().with_size(80, 0).with_label("Ok");
        style_button(&mut ok);
        pack.end();
        win.end();
        win.make_modal(true);
        win.show();
        ok.set_callback({
            let mut win = win.clone();
            move |_| {
                win.hide();
            }
        });
        while win.shown() {
            app::wait();
        }
        Self { window: win, input }
    }
    pub fn value(&self) -> String {
        self.input.value()
    }
}


fn main() {
    let a = app::App::default();
    app::set_font(Font::Times);
    app::set_screen_scale(0, 1.2);
    let mut win = window::Window::default()
        .with_size(600, 400)
        .with_label("My App");
    win.set_color(Color::from_rgb(240, 240, 240));
    let mut btn = button::Button::default()
        .with_size(80, 30)
        .with_label("Click")
        .center_of_parent();
    style_button(&mut btn);
    let mut frame = frame::Frame::new(
        btn.x() - 40,
        btn.y() - 100,
        btn.w() + 80,
        30,
        None,
    );
    frame.set_frame(FrameType::BorderBox);
    frame.set_color(Color::Red.inactive());
    win.end();
    win.show();
    btn.set_callback(move |_| {
        let dlg = show_dialog("Whatever title!");
        frame.set_label(&dlg.value());
        app::delete_widget(dlg.window);
    });
    a.run().unwrap();
}
```

Here it just takes the title of the dialog

**Comment by mark-summerfield** [comment]
I'm very sorry for that silly mistake!

Here's possibly another one. Out of interest and because I think it would be cleaner I tried implementing the Drop trait (and of course remove the delete_widget() call from the callback):

```
impl Drop for MyDialog {
    fn drop(&mut self) {
        println!("### Dropping MyDialog");
        //app::delete_widget(self.window);
        window::Window::delete(self.window);
    }
}
```
Neither of these works. I also tried &self.window and &mut self.window but without success.

I'm also reading through and trying out the rest of your examples, so may encounter a solution in the process.

**Comment by MoAlyousef** [comment]
You can use self.window.clone()

This doens't actually clone the window

**Comment by mark-summerfield** [comment]
Ah, I hadn't realised about clone() not cloning! Here's the code I ended up using:
```
impl Drop for MyDialog {
    fn drop(&mut self) {
        app::delete_widget(self.window.clone());
    }
}
```
It is much nicer than needing to explicitly delete (& I tested it with a println!() first).

Oh, and I tried out dialog::message_title() and dialog::message_title_default() and they both worked fine.

PS My rust is a bit rusty and this is the first time I've used FLTK (I've used Qt, wxWidgets, Tk, and others, so am used to GUI toolkits). I really like how I can build with the same source on Windows & Linux and how _small_ the executables are.
